src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
> As an extensibility pattern, it needs to be core to Orleans

Activation services, including lifetime, yes. Binding to constructor parameters so that we can use attributes in them, IMO no, it doesn't need to be core to Orleans, supporting it is a nice to have.

> > "IMO still needs more thought with regards to 3rd party containers"

> Agreed, and this is subject to change. However, whatever we come up with should be the expected and supported pattern.

But that's the point, you are coupling this to GrainTypeData, which has nothing to do with how we use DI to resolve stuff.

>> "I'd couple it to DefaultGrainActivator"

> I think we should think on this.

You were quoting me out of context... My recommendation is NOT to couple it to DefaultGrainActivator or to anything. If it has to be one, I was suggesting that's at least related to DI resolution, but I really think it should not be related to anything... if you resolve this activation service from the default configured container, then fine, it looks at the attribute information by using this workaround/hack. If you are using Autofac, you can probably change the service registration to something that is more straightforward that just looks at the metadata without having to use the BindToConstructorParameter stuff at all, which is just a workaround for how limited the default container implementation from Microsoft.Extensions.DependencyInjection is.

In summary, and being concrete with the IStorageFacet example, when I mentioned that 3rd party container integration needs more thought, I'm alluding at whether we want to also provide an alternative IStorageFacet implementation that doesn't do the binding to a constructor via the AttributedStorageFacet, but use a more formal approach for that container that is less hacky. The storage implementation doesn't change at all, just how it's wired up. Sure, Autofac can still use the hacky approach by if there is not an explicit override implementation for that container. After all, the reason you decide to use a different container is for no other reason that you prefer to configure things in a different idiomatic way.
That's also the reason why I don't think the storage interface should have the *Facet* suffix at all, because I don't want the resolution and binding to a constructor parameter to pollute the actual value of the real implementation of the storage feature. You could also configure your container in the ConfigureServices method to define the metadata the storage facet should use, people should not be forced to go through an attribute and define it statically in the Type.